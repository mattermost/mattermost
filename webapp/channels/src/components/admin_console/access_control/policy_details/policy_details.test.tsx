// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {shallow} from 'enzyme';
import React from 'react';
import {act} from 'react-dom/test-utils';

import type {ChannelWithTeamData} from '@mattermost/types/channels';

import PolicyDetails from './policy_details';

jest.mock('utils/browser_history', () => ({
    getHistory: () => ({
        push: jest.fn(),
    }),
}));

describe('components/admin_console/access_control/policy_details/PolicyDetails', () => {
    const mockCreatePolicy = jest.fn();
    const mockUpdatePolicy = jest.fn();
    const mockDeletePolicy = jest.fn();
    const mockSearchChannels = jest.fn();
    const mockSetChannelListSearch = jest.fn();
    const mockSetChannelListFilters = jest.fn();
    const mockOnRemoveCallback = jest.fn();
    const mockOnUndoRemoveCallback = jest.fn();
    const mockOnAddCallback = jest.fn();
    const mockFetchPolicy = jest.fn();
    const mockSetNavigationBlocked = jest.fn();
    const mockAssignChannelsToAccessControlPolicy = jest.fn();
    const mockUnassignChannelsFromAccessControlPolicy = jest.fn();
    const mockGetAccessControlExpressionAutocomplete = jest.fn();
    const mockGetAccessControlFields = jest.fn();
    const mockCreateJob = jest.fn();
    const mockUpdateAccessControlPolicyActive = jest.fn();

    const defaultProps = {
        policyId: 'policy1',
        channels: [
            {id: 'channel1', name: 'Channel 1', display_name: 'Channel 1', team_display_name: 'Team 1', type: 'O'} as ChannelWithTeamData,
            {id: 'channel2', name: 'channel2', display_name: 'Channel 2', team_display_name: 'Team 2', type: 'P'} as ChannelWithTeamData,
        ],
        totalCount: 2,
        searchTerm: '',
        filters: {},
        onRemoveCallback: mockOnRemoveCallback,
        onUndoRemoveCallback: mockOnUndoRemoveCallback,
        onAddCallback: mockOnAddCallback,
        channelsToRemove: {},
        channelsToAdd: {},
        autocompleteResult: {entities: {}},
        actions: {
            createPolicy: mockCreatePolicy,
            updatePolicy: mockUpdatePolicy,
            deletePolicy: mockDeletePolicy,
            searchChannels: mockSearchChannels,
            setChannelListSearch: mockSetChannelListSearch,
            setChannelListFilters: mockSetChannelListFilters,
            fetchPolicy: mockFetchPolicy,
            setNavigationBlocked: mockSetNavigationBlocked,
            assignChannelsToAccessControlPolicy: mockAssignChannelsToAccessControlPolicy,
            unassignChannelsFromAccessControlPolicy: mockUnassignChannelsFromAccessControlPolicy,
            getAccessControlExpressionAutocomplete: mockGetAccessControlExpressionAutocomplete,
            getAccessControlFields: mockGetAccessControlFields,
            createJob: mockCreateJob,
            updateAccessControlPolicyActive: mockUpdateAccessControlPolicyActive,
        },
    };

    beforeEach(() => {
        mockCreatePolicy.mockReset();
        mockUpdatePolicy.mockReset();
        mockDeletePolicy.mockReset();
        mockSearchChannels.mockReset();
        mockSetChannelListSearch.mockReset();
        mockSetChannelListFilters.mockReset();
        mockOnRemoveCallback.mockReset();
        mockOnUndoRemoveCallback.mockReset();
        mockOnAddCallback.mockReset();
        mockFetchPolicy.mockReset();
        mockSetNavigationBlocked.mockReset();
        mockAssignChannelsToAccessControlPolicy.mockReset();
        mockUnassignChannelsFromAccessControlPolicy.mockReset();
        mockGetAccessControlExpressionAutocomplete.mockReset();
        mockGetAccessControlFields.mockReset();
        mockCreateJob.mockReset();
        mockUpdateAccessControlPolicyActive.mockReset();
    });

    test('should match snapshot with new policy', () => {
        const props = {
            ...defaultProps,
            policyId: '',
        };
        const wrapper = shallow(<PolicyDetails {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('should match snapshot with existing policy', () => {
        const props = {
            ...defaultProps,
            actions: {
                ...defaultProps.actions,
                fetchPolicy: jest.fn().mockResolvedValue({
                    data: {
                        policy: {
                            id: 'policy1',
                            name: 'Policy 1',
                            rules: [{expression: 'true'}],
                        },
                    },
                }),
            },
        };
        const wrapper = shallow(<PolicyDetails {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('should handle delete policy', async () => {
        const props = {
            ...defaultProps,
            policyId: 'policy1',
            actions: {
                ...defaultProps.actions,
                deletePolicy: mockDeletePolicy.mockResolvedValue({data: {}}),
            },
        };

        const wrapper = shallow(<PolicyDetails {...props}/>);

        // Find the delete-policy card
        const deleteCard = wrapper.find('Card.delete-policy');
        expect(deleteCard.exists()).toBe(true);

        // Find the header with button inside the card
        const deleteButtonHeader = deleteCard.find('TitleAndButtonCardHeader');
        expect(deleteButtonHeader.exists()).toBe(true);

        const onClickProp = deleteButtonHeader.props().onClick;
        expect(onClickProp).toBeDefined();

        // Click the delete button
        await act(async () => {
            if (onClickProp) {
                await onClickProp({} as React.MouseEvent);
            }
        });

        // Update the wrapper to see the modal
        wrapper.update();

        // Find the confirmation modal
        const confirmationModal = wrapper.find('GenericModal');
        expect(confirmationModal.exists()).toBe(true);

        // Trigger the confirmation
        await act(async () => {
            // Get the handleConfirm prop with proper type assertion
            const handleConfirm = (confirmationModal.props() as any).handleConfirm;
            if (handleConfirm) {
                await handleConfirm();
            }
        });

        expect(mockDeletePolicy).toHaveBeenCalledWith('policy1');
    });
});
