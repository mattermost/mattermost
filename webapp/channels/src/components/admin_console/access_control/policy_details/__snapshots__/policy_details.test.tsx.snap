// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/admin_console/access_control/policy_details/PolicyDetails should match snapshot with existing policy 1`] = `
<div
  className="wrapper--fixed AccessControlPolicySettings"
>
  <AdminHeader
    withBackButton={true}
  >
    <div>
      <Connect(Component)
        className="fa fa-angle-left back"
        to="/admin_console/user_management/attribute_based_access_control"
      />
      <MemoizedFormattedMessage
        defaultMessage="Edit Access Control Policy"
        id="admin.access_control.policy.edit_policy.title"
      />
    </div>
  </AdminHeader>
  <div
    className="admin-console__wrapper"
  >
    <div
      className="admin-console__content"
    >
      <div
        className="admin-console__setting-group"
      >
        <TextSetting
          id="admin.access_control.policy.edit_policy.policyName"
          inputClassName="col-sm-8"
          label={
            <Memo(MemoizedFormattedMessage)
              defaultMessage="Access control policy name:"
              id="admin.access_control.policy.edit_policy.policyName"
            />
          }
          labelClassName="col-sm-4 vertically-centered-label"
          onChange={[Function]}
          placeholder="Add a unique policy name"
          value=""
        />
        <Memo(BooleanSetting)
          helpText={
            <Memo(MemoizedFormattedMessage)
              defaultMessage="Users who match the attribute values configured below will be automatically added as new members. Regardless of this setting, users who later no longer match the configured attribute values will be removed from the channel after the next sync."
              id="admin.access_control.policy.edit_policy.autoSyncMembership.description"
            />
          }
          id="admin.access_control.policy.edit_policy.autoSyncMembership"
          label={
            <div
              className="vertically-centered-label"
            >
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Auto-add members based on access rules:"
                id="admin.access_control.policy.edit_policy.autoSyncMembership"
              />
            </div>
          }
          onChange={[Function]}
          setByEnv={false}
          value={false}
        />
      </div>
      <Card
        className="console"
        expanded={true}
      >
        <CardHeader>
          <TitleAndButtonCardHeader
            buttonText={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Switch to Advanced Mode"
                id="admin.access_control.policy.edit_policy.switch_to_advanced"
              />
            }
            isDisabled={false}
            onClick={[Function]}
            subtitle="Select user attributes and values as rules to restrict channel membership."
            title="Attribute based access rules"
          />
        </CardHeader>
        <CardBody>
          <TableEditor
            onChange={[Function]}
            onValidate={[Function]}
            userAttributes={Array []}
            value=""
          />
        </CardBody>
      </Card>
      <Card
        className="console add-channels"
        expanded={true}
      >
        <CardHeader>
          <TitleAndButtonCardHeader
            buttonText={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Add channels"
                id="admin.access_control.policy.edit_policy.channel_selector.addChannels"
              />
            }
            onClick={[Function]}
            subtitle={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Add channels that this property based access policy will apply to."
                id="admin.access_control.policy.edit_policy.channel_selector.subtitle"
              />
            }
            title={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Assigned channels"
                id="admin.access_control.policy.edit_policy.channel_selector.title"
              />
            }
          />
        </CardHeader>
        <CardBody
          expanded={true}
        >
          <Connect(ChannelList)
            channelsToAdd={Object {}}
            channelsToRemove={Object {}}
            onRemoveCallback={[Function]}
            onUndoRemoveCallback={[Function]}
            policyId="policy1"
          />
        </CardBody>
      </Card>
      <Card
        className="console delete-policy"
        expanded={true}
      >
        <CardHeader>
          <TitleAndButtonCardHeader
            buttonText={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Delete"
                id="admin.access_control.policy.edit_policy.delete_policy.delete"
              />
            }
            isDisabled={false}
            onClick={[Function]}
            subtitle={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="This policy will be deleted and cannot be recovered."
                id="admin.access_control.policy.edit_policy.delete_policy.subtitle"
              />
            }
            title={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Delete policy"
                id="admin.access_control.policy.edit_policy.delete_policy.title"
              />
            }
          />
        </CardHeader>
      </Card>
    </div>
  </div>
  <div
    className="admin-console-save"
  >
    <SaveButton
      defaultMessage={
        <Memo(MemoizedFormattedMessage)
          defaultMessage="Save"
          id="admin.access_control.edit_policy.save"
        />
      }
      disabled={true}
      onClick={[Function]}
    />
    <Connect(Component)
      className="btn btn-quaternary"
      to="/admin_console/user_management/attribute_based_access_control"
    >
      <MemoizedFormattedMessage
        defaultMessage="Cancel"
        id="admin.access_control.edit_policy.cancel"
      />
    </Connect(Component)>
  </div>
</div>
`;

exports[`components/admin_console/access_control/policy_details/PolicyDetails should match snapshot with new policy 1`] = `
<div
  className="wrapper--fixed AccessControlPolicySettings"
>
  <AdminHeader
    withBackButton={true}
  >
    <div>
      <Connect(Component)
        className="fa fa-angle-left back"
        to="/admin_console/user_management/attribute_based_access_control"
      />
      <MemoizedFormattedMessage
        defaultMessage="Edit Access Control Policy"
        id="admin.access_control.policy.edit_policy.title"
      />
    </div>
  </AdminHeader>
  <div
    className="admin-console__wrapper"
  >
    <div
      className="admin-console__content"
    >
      <div
        className="admin-console__setting-group"
      >
        <TextSetting
          id="admin.access_control.policy.edit_policy.policyName"
          inputClassName="col-sm-8"
          label={
            <Memo(MemoizedFormattedMessage)
              defaultMessage="Access control policy name:"
              id="admin.access_control.policy.edit_policy.policyName"
            />
          }
          labelClassName="col-sm-4 vertically-centered-label"
          onChange={[Function]}
          placeholder="Add a unique policy name"
          value=""
        />
        <Memo(BooleanSetting)
          helpText={
            <Memo(MemoizedFormattedMessage)
              defaultMessage="Users who match the attribute values configured below will be automatically added as new members. Regardless of this setting, users who later no longer match the configured attribute values will be removed from the channel after the next sync."
              id="admin.access_control.policy.edit_policy.autoSyncMembership.description"
            />
          }
          id="admin.access_control.policy.edit_policy.autoSyncMembership"
          label={
            <div
              className="vertically-centered-label"
            >
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Auto-add members based on access rules:"
                id="admin.access_control.policy.edit_policy.autoSyncMembership"
              />
            </div>
          }
          onChange={[Function]}
          setByEnv={false}
          value={false}
        />
      </div>
      <Card
        className="console"
        expanded={true}
      >
        <CardHeader>
          <TitleAndButtonCardHeader
            buttonText={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Switch to Advanced Mode"
                id="admin.access_control.policy.edit_policy.switch_to_advanced"
              />
            }
            isDisabled={false}
            onClick={[Function]}
            subtitle="Select user attributes and values as rules to restrict channel membership."
            title="Attribute based access rules"
          />
        </CardHeader>
        <CardBody>
          <TableEditor
            onChange={[Function]}
            onValidate={[Function]}
            userAttributes={Array []}
            value=""
          />
        </CardBody>
      </Card>
      <Card
        className="console add-channels"
        expanded={true}
      >
        <CardHeader>
          <TitleAndButtonCardHeader
            buttonText={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Add channels"
                id="admin.access_control.policy.edit_policy.channel_selector.addChannels"
              />
            }
            onClick={[Function]}
            subtitle={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Add channels that this property based access policy will apply to."
                id="admin.access_control.policy.edit_policy.channel_selector.subtitle"
              />
            }
            title={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Assigned channels"
                id="admin.access_control.policy.edit_policy.channel_selector.title"
              />
            }
          />
        </CardHeader>
        <CardBody
          expanded={true}
        >
          <Connect(ChannelList)
            channelsToAdd={Object {}}
            channelsToRemove={Object {}}
            onRemoveCallback={[Function]}
            onUndoRemoveCallback={[Function]}
            policyId=""
          />
        </CardBody>
      </Card>
    </div>
  </div>
  <div
    className="admin-console-save"
  >
    <SaveButton
      defaultMessage={
        <Memo(MemoizedFormattedMessage)
          defaultMessage="Save"
          id="admin.access_control.edit_policy.save"
        />
      }
      disabled={true}
      onClick={[Function]}
    />
    <Connect(Component)
      className="btn btn-quaternary"
      to="/admin_console/user_management/attribute_based_access_control"
    >
      <MemoizedFormattedMessage
        defaultMessage="Cancel"
        id="admin.access_control.edit_policy.cancel"
      />
    </Connect(Component)>
  </div>
</div>
`;
